lib/src/simple_config_origin.cc

```diff
@@ -54,6 +57,171 @@ namespace hocon {
+                auto a = dynamic_pointer_cast<const simple_config_origin>(remaining.back());
+                remaining.pop_back();
+
+                shared_origin merged = merge_three(a, b, c);
+                remaining.push_back(merged);
+            }
+
+            //  should be down to either 1 or 2
+            return merge_origins(remaining);
+        }
+    }
+
+    shared_ptr<const simple_config_origin> simple_config_origin::merge_two(shared_ptr<const simple_config_origin> a,
+                                                                           shared_ptr<const simple_config_origin> b) {
+        string merged_desc;
+        int merged_start_line;
+        int merged_end_line;
+
+        origin_type merged_type;
+        if (a->_origin_type == b->_origin_type) {
+            merged_type = a->_origin_type;
+        } else {
+            merged_type = origin_type::GENERIC;
+        }
+
+        const string MERGE_OF_PREFIX { "merge of " };
+        // first use the "description" field which has no line numbers
+        // cluttering it.
+        string a_desc { a->description() };
+        string b_desc { b->description() };
+        if (boost::starts_with(a_desc, MERGE_OF_PREFIX)) {
+            a_desc = a_desc.substr(MERGE_OF_PREFIX.length());
+        }
+
+        if (boost::starts_with(b_desc, MERGE_OF_PREFIX)) {
+            b_desc = b_desc.substr(MERGE_OF_PREFIX.length());
+        }
+
+        if (a_desc == b_desc) {
+            merged_desc = a_desc;
+
+            if (a->line_number() < 0) {
+                merged_start_line = b->line_number();
+            } else if (b->line_number() < 0) {
+                merged_start_line = a->line_number();
+            } else {
+                merged_start_line = max(a->_end_line_number, b->_end_line_number);
+            }
+
+            merged_end_line = max(a->_end_line_number, b->_end_line_number);
+        } else {
+            // this whole merge song-and-dance was intended to avoid this case
+            // whenever possible, but we've lost. Now we have to lose some
+            // structured information and cram into a string.
+
+            // description() method includes line numbers, so use it instead
+            // of description field.
+            string a_full = a->description();
+            string b_full = b->description();
+            if (boost::starts_with(a_full, MERGE_OF_PREFIX)) {
+                a_full = a_full.substr(MERGE_OF_PREFIX.length());
+            }
+
+            if (boost::starts_with(b_full, MERGE_OF_PREFIX)) {
+                b_full = b_full.substr(MERGE_OF_PREFIX.length());
+            }
+
+            merged_desc = MERGE_OF_PREFIX + a_full + "," + b_full;
+            merged_start_line = -1;
+            merged_end_line = -1;
+        }
+
+        string merged_resource;
+        if (a->_resource_or_null == b->_resource_or_null) {
+            merged_resource = a->_resource_or_null;
+        } else {
+            merged_resource = "";
```

Comment:
Can remove the else clause, that's the default initialized behavior for merged_resource.